File: project/6.02_継承と変数/java/java_ex6_02_1/src/java_ex6_02_1/Ex6_02_1.java
package java_ex6_02_1;

class GrandParent {
	int i;
	double d;
}

class Parent extends GrandParent {
	float f;
	String d;
}

class Child extends Parent {
	String s;
}

class M {
	float ff;
	String ss;
}

class N extends M {
	Double dd;
}

public class Ex6_02_1 {

	public static void main(String[] args) {
		Parent p;				// 親は子を受け入れられる
		p = new Child();
		p = new Parent();
		
		Child c;
		c = new Child();
//		c = new Parent();		// 子は親を受け入れられない

		c.i = 10;				// 子は継承元の変数をすべて参照できる(資産を継承するイメージ)
		c.f = 2.3f;
		c.s = "4567";
		c.d = "";
		// ただし同じ変数名が存在する場合は親の情報が隠ぺいされる(子の情報で上書きされるイメージ)
		// c.d はString型のdしか見えない
		System.out.println(c.d.getClass().getSimpleName());	// String
		
		N n = new N();
		n.dd = 1.1;
		n.ff = 2.2f;
		n.ss = "3.3";
		System.out.println( n.dd );
		System.out.println( n.ff );
		System.out.println( n.ss );
		
	}

}

File: project/6.02_継承と変数/py/py_ex6_02_1/src/ex6_02_1.py
'''
Created on 2023/05/06

@author: Keroichi.T
'''

class GrandParent( object ):
    def __init__(self):
        self.i = 10
        self.d = 1
        
class Parent( GrandParent ):
    def __init__(self):
        # super().__init__()
        self.f = 2.3
        self.d = "1.1"
        
class Child( Parent ):
    def __init__(self):
        super().__init__()
        self.f = "2.3"      # 同じ名前なので親の情報は見えなくなる
        self.s = "4567"



def main():
    c = Child()

    # インスタンス変数が生成されていなければ参照できない
    # サブクラスでコンストラクタの実行が必要
    # print(c.i)

    print( c.s )    # 継承元のインスタンス変数も参照可能(コンストラクタ要実行)
    print( c.d )
    # 同じインスタンス変数名がある場合は、子の情報で上書きする
    print( "{}, {}".format( c.f, type( c.f ) ) )
    
    p = Parent()
    # ↓は参照できているように見えているが、実際は違う。
    # iというインスタンス変数を新たに生成しているだけで、
    # 親クラスのインスタンス変数とはまったく別物
    p.i = 100
    print( p.i )
    

if __name__ == '__main__':
    main()